src/imitation/rewards/serialize.py

```diff
@@ -0,0 +1,99 @@
+"""Load serialized policies of different types."""
+
+from typing import Callable
+
+import numpy as np
+from stable_baselines.common.vec_env import VecEnv
+
+from imitation.rewards import discrim_net, reward_net
+from imitation.util import registry, util
+
+RewardFn = Callable[[np.ndarray, np.ndarray, np.ndarray], np.ndarray]
+RewardLoaderFn = Callable[[str, VecEnv], RewardFn]
+RewardNetLoaderFn = Callable[[str, VecEnv], reward_net.RewardNet]
+
+reward_net_registry: registry.Registry[RewardNetLoaderFn] = registry.Registry()
+reward_registry: registry.Registry[RewardLoaderFn] = registry.Registry()
+
+
+def _add_reward_net_loaders(classes):
+  for name, cls in classes.items():
+    loader = registry.build_loader_fn_require_path(cls.load)
+    reward_net_registry.register(key=name, value=loader)
+
+
+REWARD_NETS = {
+    "BasicRewardNet": reward_net.BasicRewardNet,
+    "BasicShapedRewardNet": reward_net.BasicShapedRewardNet,
+}
+_add_reward_net_loaders(REWARD_NETS)
+
+
+def _load_discrim_net(cls):
+  def f(path: str, venv: VecEnv):
+    discrim_net = cls.load(path)
+    # TODO(gleave): expose train reward as well? (hard due to action probs?)
+    return discrim_net.reward_test
+  return f
+
+
+def _add_discrim_net_loaders(classes):
+  for name, cls in classes.items():
+    reward_registry.register(key=name, value=_load_discrim_net(cls))
+
+
+_add_discrim_net_loaders({
+    "DiscrimNetAIRL": discrim_net.DiscrimNetAIRL,
+    "DiscrimNetGAIL": discrim_net.DiscrimNetGAIL,
+})
+
+
+def load_reward_net_as_fn(path: str, env: VecEnv) -> RewardFn:
+  reward_type, shaped, reward_path = path.split(':')
+  reward_net_loader = reward_net_registry.get(reward_type)
+  assert shaped in ["True", "False"]
+  shaped = shaped == "True"
+
+  # TODO(adam): leaks session
```

Comment:
I'd be OK with this, but it feels just like punting the problem elsewhere. At the current callsite in scripts/expert_demos.py:105 it seems like most the code needs the reward model. Probably we'd want to deallocate the session when the `RewardVecEnvWrapper` gets `close`(d)?

Reply:
Agree that `reward_fn` and its Session should be closed once `venv = RewardVecEnvWrapper(venv, reward_fn)` is no longer used (nearly at the end of the function).

Reply:
(`RewardVecWrapper` wouldn't make a good contextmanager because it isn't guaranteed to hold a `Session` (and wouldn't have access to reward_fn's Session anyways).)

Reply:
Conditionally setting up a context from `reward_fn` seems hard to do via a vanilla context + a no-op context switch.

https://github.com/HumanCompatibleAI/imitation/blob/2d1f29fb636b3891e8ce0c7c3845976f2df514fd/src/imitation/scripts/expert_demos.py#L99-L110

Reply:
So this might be a good use case for contextlib.ExitStack.

Something like
```
with ExitStack() as exit_stack:
  ...
  if reward_type is not None:
    reward_fn, resources = load_reward(...)   # type: Tuple[RewardFn, List["Implements .close()"]]
    for resource in resources: 
      exit_stack.push(resource)
    venv = RewardVecEnvWrapper(venv, reward_fn)
```

Reply:
Thanks ! I'd never seen `ExitStack`. I'm going to address this in a separate PR since I want to change things for `policies/serialize.py` as well.